File: src/code-samples/deepagents/forked-subagents-compiled.py
"""Forked subagents page code samples: CompiledSubAgent."""

from __future__ import annotations

# :snippet-start: forked-subagents-compiled-py
from deepagents import CompiledSubAgent, create_deep_agent
from langchain.agents import create_agent


def read_logs(path: str) -> str:
    """Read a log file."""
    return f"logs from {path}"


# A prebuilt graph with its own system prompt -- forking inherits the
# parent's message history only, not its system prompt.
incident_graph = create_agent(
    # KEEP MODEL
    model="claude-sonnet-4-6",
    tools=[read_logs],
    system_prompt="You are an incident postmortem writer.",
)

incident_responder = CompiledSubAgent(
    name="incident-responder",
    description="Continues an in-progress incident investigation and drafts the postmortem",
    runnable=incident_graph,
    mode="fork",
)

agent = create_deep_agent(
    model="claude-sonnet-4-6",
    tools=[read_logs],
    subagents=[incident_responder],
)
# :snippet-end:

# :remove-start:
assert agent is not None
# :remove-end:

File: src/code-samples/deepagents/forked-subagents-compiled.ts
// :snippet-start: forked-subagents-compiled-js
import { CompiledSubAgent, createDeepAgent } from "deepagents";
import { createAgent, tool } from "langchain";
import { z } from "zod";

const readLogs = tool(
  async ({ path }: { path: string }) => `logs from ${path}`,
  {
    name: "read_logs",
    description: "Read a log file",
    schema: z.object({ path: z.string() }),
  },
);

// A prebuilt graph with its own system prompt -- forking inherits the
// parent's message history only, not its system prompt.
const incidentGraph = createAgent({
  // KEEP MODEL
  model: "claude-sonnet-4-6",
  tools: [readLogs],
  prompt: "You are an incident postmortem writer.",
});

const incidentResponder: CompiledSubAgent = {
  name: "incident-responder",
  description: "Continues an in-progress incident investigation and drafts the postmortem",
  runnable: incidentGraph,
  mode: "fork",
};

const agent = await createDeepAgent({
  // KEEP MODEL
  model: "claude-sonnet-4-6",
  tools: [readLogs],
  subagents: [incidentResponder],
});
// :snippet-end:

// :remove-start:
if (!agent) throw new Error("agent not created");
console.log("✓ forked-subagents-compiled validated");
// :remove-end:
